File: Baekjoon/Class3/02.1620 나는야 포켓몬 마스터 이다솜.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

class Pokedex {
private:
    int _size;
    vector<string> _index_to_name;
    unordered_map<string, int> _name_to_index;
public:
    Pokedex(int reserve_size) {
        _size = 0;
        _index_to_name.reserve(reserve_size + 1);
        _index_to_name.push_back("Dummy"); // 0번 인덱스 더미 데이터
        _name_to_index.reserve(reserve_size);
    }
    void add(const string& name) {
        _size++;
        _index_to_name.push_back(name);
        _name_to_index.insert({ name, _size });
    }
    const string& get_name(int index) const {
        return _index_to_name[index];
    }
    int get_index(const string& name) const {
        return _name_to_index.at(name);
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    Pokedex pokedex(n);

    while (n--) { // 이름 등록
        string name;
        cin >> name;
        pokedex.add(name);
    }

    while (m--) { // 검색
        string target;
        cin >> target;
        if (isdigit(target[0])) { // 인덱스 검색
            int index = stoi(target);
            cout << pokedex.get_name(index) << '\n';
        }
        else // 이름 검색
            cout << pokedex.get_index(target) << '\n';
    }
    return 0;
}

File: Baekjoon/Class3/02.1620 나는야 포켓몬 마스터 이다솜.java
import java.io.*;
import java.util.HashMap;
import java.util.StringTokenizer;

class Main {
    private static class Pokedex {
        private int size;
        private String[] indexToName;
        private HashMap<String, Integer> nameToIndex;

        public Pokedex(int reserveSize) {
            size = 0;
            indexToName = new String[reserveSize + 1];
            indexToName[0] = "Dummy"; // 0번 인덱스 더미 데이터
            nameToIndex = new HashMap<>(reserveSize);
        }
        public void add(String name) {
            size++;
            indexToName[size] = name;
            nameToIndex.put(name, size);
        }
        public String getName(int index) {
            return indexToName[index];
        }
        public int getIndex(String name) {
            return nameToIndex.get(name);
        }
    }

    public static void main(String[] args) throws IOException {
        int n, m;
        Pokedex pokedex;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        pokedex = new Pokedex(n);

        for(int i = 0; i < n; i++) { // 이름 등록
            String name = br.readLine();
            pokedex.add(name);
        }

        for(int i = 0; i < m; i++) { // 검색
            String target = br.readLine();
            if(Character.isDigit(target.charAt(0))) { // 인덱스 검색
                int index = Integer.parseInt(target);
                sb.append(pokedex.getName(index)).append('\n');
            }
            else // 이름 검색
                sb.append(pokedex.getIndex(target)).append('\n');
        }
        System.out.print(sb);
        br.close();
    }
}